proposed/2022/vulnerabilities-in-restore.md

```diff
@@ -0,0 +1,391 @@
+    <NoWarn>$(NoWarn);NU1901</NoWarn>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="PackageId" Version="3.0.0" NoWarn="NU1901" />
+  </ItemGroup>
+
+```
+
+We can extend a similar functionality to the advisory urls via their own dedicated property.
+
+```xml
+<RestoreSuppressedAdvisoryUrls>https://github.com/advisories/GHSA-g3q9-xf95-8hp5</RestoreSuppressedAdvisoryUrls>
+
+ <PropertyGroup>
+    <RestoreSuppressedAdvisoryUrls>$(NoWarn);https://github.com/advisories/GHSA-g3q9-xf95-8hp5</RestoreSuppressedAdvisoryUrls>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="PackageId" Version="3.0.0" SuppressedAdvisoryUrls="https://github.com/advisories/GHSA-g3q9-xf95-8hp5" />
+  </ItemGroup>
+```
+
+The project level and package level metadata are not necessarily a package deal, we can choose to only add one.
+
+### Allow sources to upstream vulnerability information
+
+While the NuGet tooling does support multiple sources and a means to control which packages come from which source through [Package Source Mapping](https://learn.microsoft.com/en-us/nuget/consume-packages/package-source-mapping), all configurations are not going to contain nuget.org, and thus won't easily get the data curated from the [GitHub Advisory Database](https://devblogs.microsoft.com/nuget/how-to-scan-nuget-packages-for-security-vulnerabilities/).
+
+The [V3 protocol](https://learn.microsoft.com/en-us/nuget/api/overview) and the resource architecture would allow a source to link to a resource that may not be hosted by them.
+Not every source would want to setup a pipeline for processing vulnerability information. They can instead just "point" to the nuget.org information.
+
+For example, an Azure Artifacts source could have this in their index.json:
+
+```json
+{
+  "resources" [
+
+  {
+    "@id": "https://api.nuget.org/v3/vulnerability.json",
+    "@type": "VulnerabilityInfo/6.3.0",
+    "comment": "The base data for vulnerability update periodically"
+  }
+  ]
+}
+```
+
+While the nuget.org data only contains vulnerability information about packages on nuget.org, often times an `upstream` source is setup by customers, and while they reference their own private feed, they can still download packages from nuget.org. The vulnerability information is valuable in these scenarios.
+
+If this upstream approach is taken, it is very likely that the vulnerability checking has a lot of misses, but this is an understood trade-off.
+
+It is **very important* that if the upstream sources link to the nuget.org vulnerability, the actual nuget.org urls are used wherever possible, as this would allow the client to easily deduplicate the vulnerability information.
+
+### Providing vulnerability information through a json, remotely or locally
+
+Other components may want to provide vulnerability information that's not tied to a source.
+Whatever approach we take there will be a way to represent the vulnerability information through a json file.

```

Comment:
Perhaps there could be a nuget.config for a list of vulnerability info URLs? That'd be useful if I have a single private package source that doesn't point to its upstream vulnerability info URL yet.

Reply:
Yes, it would have to be something along those lines.